### ISSUE TITLE
Database connection timeout

### ISSUE THREAD

**Issue by saikris12** [atomikos/transactions-essentials#101]
We are using Atomikos transaction manager with Oracle 19c database as part of a Spring boot app that runs inside docker containers. We are observing that every now and then the application stalls with the below warnings. The processes inside the container increase abnormally and it just does not respond. 

WARN  [AbstractXPooledConnection, Atomikos:2] atomikosXAPooledConnection with sessionHandleState (1 context(s), isTerminated = false) for resource atomikosDataSource: reaping connection (and starting to collect stack traces for next reap)...

WARN  [ActiveStateHandler, Atomikos:4557] Transaction 10.200.1.4.tm157616595296710936 has timed out and will rollback.

We are using atomikos 5.0.3. 

What does this error actually mean and if the application lost connectivity to DB why is it not able to reestablish the connection and I do not understand why there is a spike in the overall processes running inside the container.

**Comment by GuyPardon** [comment]
Do you also have pool exhaustion warnings? Might be a network issue.

If you're open to a free trial of ExtremeTransactions then we can help you. Are you interested in that?

**Comment by saikris12** [comment]
No. We do not have any pool exhaustion warnings. The only warnings we see in the logs are the ones I added in my comment above. The app runs in 4 containers on 4 different VMs and the issue popped up in one of the VMs. We manually restarted the container and it was fine since then. We are connecting to Oracle database which was upgraded recently to Oracle 19c. We are using an older version of JDBC driver that was originally used for Oracle 11g.  Do you think there could be some issue because of this?

**Comment by saikris12** [comment]
A more comprehensive stack trace when the issue happened : 

`java.lang.Thread:run: 748(execution time: 0 ms of 306853 ms total, wait time: 0 ms of 306749 ms total)
  com.atomikos.timing.PooledAlarmTimer:run: 70(execution time: 0 ms of 306853 ms total, wait time: 0 ms of 306749 ms total)
    com.atomikos.timing.PooledAlarmTimer:doWait: 104(execution time: 0 ms of 59020 ms total, wait time: 0 ms of 59020 ms total)
      java.lang.Object:wait: Unknown (execution time: 59020 ms of 59020 ms total, wait time: 59020 ms of 59020 ms total)
    com.atomikos.timing.PooledAlarmTimer:notifyListeners: 94(execution time: 0 ms of 51 ms total)
      com.atomikos.icatch.imp.CoordinatorImp:alarm: 647(execution time: 0 ms of 51 ms total)
        com.atomikos.icatch.imp.ActiveStateHandler:onTimeout: 75(execution time: 0 ms of 51 ms total)
          com.atomikos.icatch.imp.CoordinatorImp:timedout: 775(execution time: 0 ms of 51 ms total)
            com.atomikos.icatch.imp.CoordinatorImp:setRollbackOnly: 697(execution time: 0 ms of 51 ms total)
              com.atomikos.icatch.imp.RollbackOnlyParticipant:<clinit>: 31(execution time: 0 ms of 51 ms total)
                com.atomikos.logging.LoggerFactory:createLogger: 20(execution time: 0 ms of 51 ms total)
                  com.atomikos.logging.Slf4JLoggerFactoryDelegate:createLogger: 15(execution time: 0 ms of 51 ms total)
                    Constructor of com.atomikos.logging.Slf4jLogger: 20(execution time: 0 ms of 51 ms total)
                      java.net.URL:openConnection: 979(execution time: 51 ms of 51 ms total)
    com.atomikos.timing.PooledAlarmTimer:doWait: 104(execution time: 0 ms of 165127 ms total, wait time: 0 ms of 165127 ms total)
      java.lang.Object:wait: Unknown (execution time: 165127 ms of 165127 ms total, wait time: 165127 ms of 165127 ms total)
    com.atomikos.timing.PooledAlarmTimer:notifyListeners: 91(execution time: 53 ms of 53 ms total)
    com.atomikos.timing.PooledAlarmTimer:doWait: 104(execution time: 0 ms of 82602 ms total, wait time: 0 ms of 82602 ms total)
      java.lang.Object:wait: Unknown (execution time: 82602 ms of 82602 ms total, wait time: 82602 ms of 82602 ms total)`
